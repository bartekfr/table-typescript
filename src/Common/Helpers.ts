import PapaParse, { UnparseObject } from 'papaparse';
import { saveAs } from 'file-saver';

// Immutable array helpers
// from https://redux.js.org/recipes/structuringreducers/immutableupdatepatterns
export function insertItem<T = string>(array: T[], index: number, item: T) {
  const newArray: T[] = array.slice();
  newArray.splice(index, 0, item as T);
  return newArray;
}

export function removeItem<T = string>(array: T[], index: number) {
  const newArray = array.slice();
  newArray.splice(index, 1);
  return newArray;
}

export function updateItem<T>(array: T[], index: number, newItem: T) {
  return array.map((item, i) => {
    if (i !== index) {
      return item;
    }
    return newItem;
  });
}


// saves data as CSV file
export function saveAsCSV(data: UnparseObject) {
  const csvString = PapaParse.unparse(data);
  const blob = new Blob([csvString], { type: 'text/csv;charset=utf-8' });
  saveAs(blob, 'tableCSV.csv');
}
