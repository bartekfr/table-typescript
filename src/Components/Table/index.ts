export { default } from './Table';
export type { TableProps, TableState } from './Table';
