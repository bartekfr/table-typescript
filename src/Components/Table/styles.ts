import styled from 'styled-components'

const thColor2 = 'rgb(212, 212, 231)';
const thColor1 = 'rgb(230, 235, 245)';

// Table component context menu
// TODO: move nested elements to separate styled components
export const Menu = styled.div`
  display: none;
  align-content: center;
  position: absolute;

  button {
    flex: 1 1 auto;
    padding: 1px;
    border: 0 none;
    background: none;
    font-size: 14px;;

    &:nth-child(2) {
      font-size: 22px;
      font-weight: bold;
    }
  }
`

const menuWidth = `30px`;
export const MenuRow = styled(Menu)`
  flex-direction: column;
  height: calc(100% + 20px);
  top: -10px;
  width: ${menuWidth};
  left: -${menuWidth};
`

export const MenuCol = styled(Menu)`
  width: 100%;
  height: 30px;
  position: absolute;
  right: 0;
  top: -30px;
`

export const Table = styled.div`
  padding: 40px 60px;

  table {
    margin-bottom: 20px;
    border: 1px solid #000;
    border-collapse: collapse;
  }

  tr,
  td,
  th {
    position: relative;
    border: 1px solid #000;
  }

  tr {
    &:first-child {
      th:first-child {
        button:nth-child(1),
        button:nth-child(2) {
          display: none;
        }
      }
    }

    &:not(:first-child) > th:first-child {
      background: ${thColor1};
    }
  }

  td,
  th {
    min-width: 75px;
    max-width: 200px;
    height: 50px;
    background-clip: padding-box !important; // fix for dissapearing cell borders in FF
  }

  th {
    background-color: #ccc;
    background: linear-gradient(to bottom, ${thColor1} 0%, ${thColor1} 50%, ${thColor2} 100%);

    &:hover {
      background: ${thColor2};

      ${Menu} {
        display: flex;
      }
    }
  }

  td:hover {
    background: rgba(255, 255, 0, 0.2);
  }

  input {
    width: 100%;
    border: 0 none;
    padding: 15px 22px;
    background: none;
    font-size: 16px;
  }
`

export const Hint = styled.p`
  color: #f00;
  margin: 5px 0 0;
  padding: 0;
  font-size: 14px;
`


// Table component export to CSV button
export const CsvBtn = styled.button`
  padding: 10px 15px;
  margin: 20px 0;
  border: 0;
  background: #0058dc;
  color: #fff;
`
