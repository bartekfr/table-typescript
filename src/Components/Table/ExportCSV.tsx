import React from 'react';
import { UnparseObject } from 'papaparse';
import { saveAsCSV } from '../../Common/Helpers';
import { CsvBtn } from './styles'

interface ExportCSVProps {
  data: UnparseObject
}

const ExportCSV: React.FunctionComponent<ExportCSVProps> = ({ data }) => (
  <CsvBtn type="button" onClick={() => saveAsCSV(data)}>Save as CSV</CsvBtn>
);

export default ExportCSV;
