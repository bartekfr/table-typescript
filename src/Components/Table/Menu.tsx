import React from 'react';
import* as Styled from './styles'

interface MenuProps {
  add: (i: number) => void
  remove?: (i: number) => void
  index: number
}

export const MenuRow: React.FunctionComponent<MenuProps> = ({
  add,
  remove,
  index,
}) => (
  <Styled.MenuRow>
    <button title="Add row above" type="button" onClick={() => add(index)}>
      &#9650;
    </button>
    <button title="Remove row" type="button" onClick={() => remove && remove(index)}>
      &#215;
    </button>
    <button title="Add row below" type="button" onClick={() => add(index + 1)}>
      &#9660;
    </button>
  </Styled.MenuRow>
);

export const MenuCol: React.FunctionComponent<MenuProps> = ({
  add,
  remove,
  index,
}) => (
  <Styled.MenuCol>
    <button title="Add column to the left" type="button" onClick={() => add(index)}>
      &#9668;
    </button>
    <button title="Remove column" type="button" onClick={() => remove && remove(index)}>
      &#215;
    </button>
    <button title="Add column to the right" type="button" onClick={() => add(index + 1)}>
      &#9658;
    </button>
  </Styled.MenuCol>
);
