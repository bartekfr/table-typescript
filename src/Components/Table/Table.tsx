import React from 'react';
import { MenuCol, MenuRow } from './Menu';
import { updateItem, insertItem, removeItem } from '../../Common/Helpers';
import ExportCSVBtn from './ExportCSV';
import * as Styled from './styles';

const CellInput: React.FunctionComponent<{
  value: string
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}> = props => (
  <input type="text" {...props} />
);

interface TableData {
  columnNames: string[]
  cells: string[][]
}

export interface TableProps {
  onChange: (data: TableData) => void,
  data: TableData,
}

export type TableState = TableData

class Table extends React.Component<TableProps, TableState> {
  state: TableState = this.props.data

  constructor(props: TableProps) {
    super(props);
    this.updateCell = this.updateCell.bind(this);
    this.updateColName = this.updateColName.bind(this);
    this.addRow = this.addRow.bind(this);
    this.addCol = this.addCol.bind(this);
    this.removeRow = this.removeRow.bind(this);
    this.removeCol = this.removeCol.bind(this);
  }

  componentDidUpdate() {
    this.onChange();
  }

  onChange() {
    const { columnNames, cells } = this.state;
    const { onChange } = this.props;
    onChange({
      columnNames,
      cells,
    });
  }

  updateColName(i: number, e: React.ChangeEvent<HTMLInputElement>) {
    const { target: { value } } = e;
    const { columnNames } = this.state;
    const newRow = updateItem(columnNames, i, value);
    this.setState({
      columnNames: newRow,
    });
  }

  updateCell(i: number, j: number, e: React.ChangeEvent<HTMLInputElement>) {
    const { target: { value } } = e;
    const { cells } = this.state;
    const newRow = updateItem(cells[i], j, value);
    const newCells = updateItem(cells, i, newRow);
    this.setState({
      cells: newCells,
    });
  }

  addRow(i: number) {
    const { cells, columnNames } = this.state;
    const newRow: string[] = new Array(columnNames.length).fill('');
    const newCells = insertItem(cells, i, newRow);
    this.setState({
      cells: newCells,
    });
  }

  addCol(j: number) {
    const { cells, columnNames } = this.state;
    const newCells = cells.map(row => insertItem(row, j, ''));
    this.setState({
      cells: newCells,
      columnNames: insertItem(columnNames, j, `New column ${j + 1}`),
    });
  }

  removeRow(i: number) {
    const { cells } = this.state;
    const newCells = removeItem(cells, i);
    this.setState({
      cells: newCells,
    });
  }

  removeCol(i: number) {
    const { cells, columnNames } = this.state;
    const newCells = cells.map(row => removeItem(row, i));
    this.setState({
      cells: newCells,
      columnNames: removeItem(columnNames, i),
    });
  }

  render() {
    const { cells, columnNames } = this.state;
    const papaParseData = {
      fields: columnNames,
      data: cells.length ? cells : '',
    };

    // Generally array indexes shouldn't be used as keys but in this specific case
    // it shouldn't cause any issues
    return (
      <Styled.Table>
        <table>
          <tbody>
            {/* Column row - START */}
            <tr>
              <th>
                <MenuCol add={this.addCol} index={-1} />
                <MenuRow index={-1} add={this.addRow} />
                id
              </th>
              {
                columnNames.map((col, i) => (
                  <th key={i}>
                    <MenuCol add={this.addCol} index={i} remove={this.removeCol} />
                    <CellInput value={col} onChange={e => this.updateColName(i, e)} />
                  </th>
                ))
              }
            </tr>
            {/* Column row - END */}
            {/* Data Row  - START */}
            {
              cells.map((row, i) => (
                <tr key={i}>
                  <th>
                    <MenuRow add={this.addRow} index={i} remove={this.removeRow} />
                    {i + 1}
                  </th>
                  {row.map((_cell, j) => (
                    <td key={j}>
                      <CellInput value={cells[i][j]} onChange={e => this.updateCell(i, j, e)} />
                    </td>
                  ))}
                </tr>
              ))
            }
            {/* Data Row  - END */}
          </tbody>
        </table>
        <Styled.Hint>{!columnNames.length ? 'No columns defined.' : null }</Styled.Hint>
        <Styled.Hint>{!cells.length ? 'No rows defined.' : null }</Styled.Hint>
        <ExportCSVBtn data={papaParseData} />
      </Styled.Table>
    );
  }
}

export default Table;
