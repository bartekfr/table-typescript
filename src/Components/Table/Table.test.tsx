import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Table, { TableProps, TableState } from './index';
import * as Helpers from '../../Common/Helpers'
import * as Styled from './styles'

const data = {
  columnNames: ['col1', 'col2', 'col3'],
  cells: [
    ['a1', 'a2', 'a3'],
    ['b1', 'b2', 'b3'],
    ['c1', 'c2', 'c3'],
  ],
};

describe('Table component', () => {
  let wrapper: ShallowWrapper<TableProps, TableState, Table>;
  let instance: Table
  const onChange = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<Table onChange={onChange} data={data} />);
    instance = wrapper.instance();
    jest.spyOn(Helpers, 'insertItem')
    jest.spyOn(Helpers, 'removeItem')
    jest.spyOn(Helpers, 'updateItem')
    jest.clearAllMocks()
  });

  it('adds row even if there are no columns', () => {
    const localWrapper: ShallowWrapper<TableProps, TableState, Table> = shallow(<Table onChange={onChange} data={{
      columnNames: [],
      cells: []
    }} />);
    const localInstance = localWrapper.instance();
    localInstance.addRow(0);
    expect(Helpers.insertItem).toHaveBeenCalledTimes(1)
    expect(Helpers.insertItem).toHaveBeenLastCalledWith([], 0, [])
    // wrapper.update();
    expect(localWrapper.state('cells')).toEqual([
      [],
    ]);
    localInstance.addRow(0);
    // wrapper.update();
    expect(Helpers.insertItem).toHaveBeenCalledTimes(2)
    expect(Helpers.insertItem).toHaveBeenLastCalledWith([[]], 0, [])
    expect(localWrapper.state('cells')).toEqual([
      [],
      [],
    ]);
  });

  it('adds row', () => {
    // wrapper.update();
    instance.addRow(1);
    // wrapper.update();
    expect(Helpers.insertItem).toHaveBeenCalledTimes(1)
    expect(Helpers.insertItem).toHaveBeenLastCalledWith(data.cells, 1, ['', '', ''])
    expect(wrapper.state('cells')).toEqual([
      ['a1', 'a2', 'a3'],
      ['', '', ''],
      ['b1', 'b2', 'b3'],
      ['c1', 'c2', 'c3'],
    ]);

    instance.addRow(3);
    // wrapper.update();
    expect(Helpers.insertItem).toHaveBeenCalledTimes(2)
    expect(wrapper.state('cells')).toEqual([
      ['a1', 'a2', 'a3'],
      ['', '', ''],
      ['b1', 'b2', 'b3'],
      ['', '', ''],
      ['c1', 'c2', 'c3'],
    ]);
  });

  it('removes row', () => {
    // wrapper.update();
    instance.removeRow(1);
    // wrapper.update();
    expect(Helpers.removeItem).toHaveBeenCalledTimes(1)
    expect(Helpers.removeItem).toHaveBeenLastCalledWith(data.cells, 1)
    expect(wrapper.state('cells')).toEqual([
      ['a1', 'a2', 'a3'],
      ['c1', 'c2', 'c3'],
    ]);

    instance.removeRow(0);
    // wrapper.update();
    expect(Helpers.removeItem).toHaveBeenCalledTimes(2)
    expect(Helpers.removeItem).toHaveBeenLastCalledWith([
      ['a1', 'a2', 'a3'],
      ['c1', 'c2', 'c3'],
    ], 0)
    expect(wrapper.state('cells')).toEqual([
      ['c1', 'c2', 'c3'],
    ]);
  });

  it('adds columns even if there are no rows', () => {
    const localWrapper: ShallowWrapper<TableProps, TableState, Table> = shallow(<Table onChange={onChange} data={{
      columnNames: [],
      cells: []
    }} />);
    const instance = localWrapper.instance();
    instance.addCol(0);
    expect(Helpers.insertItem).toHaveBeenCalledTimes(1)
    expect(Helpers.insertItem).toHaveBeenLastCalledWith([], 0, 'New column 1')
    expect(localWrapper.state('columnNames')).toEqual(['New column 1']);
    expect(localWrapper.state('cells')).toEqual([]);
  });

  it('adds columns', () => {
    // wrapper.update();
    instance.addCol(0);
    // wrapper.update();
    expect(Helpers.insertItem).toHaveBeenLastCalledWith(data.columnNames, 0, 'New column 1')
    expect(wrapper.state('cells')).toEqual([
      ['', 'a1', 'a2', 'a3'],
      ['', 'b1', 'b2', 'b3'],
      ['', 'c1', 'c2', 'c3'],
    ]);
    expect(wrapper.state('columnNames')).toEqual(['New column 1', 'col1', 'col2', 'col3']);

    instance.addCol(2);
    // wrapper.update();
    expect(Helpers.insertItem).toHaveBeenLastCalledWith(['New column 1', ...data.columnNames], 2, 'New column 3')
    expect(wrapper.state('cells')).toEqual([
      ['', 'a1', '', 'a2', 'a3'],
      ['', 'b1', '', 'b2', 'b3'],
      ['', 'c1', '', 'c2', 'c3'],
    ]);
    expect(wrapper.state('columnNames')).toEqual(['New column 1', 'col1', 'New column 3', 'col2', 'col3']);
  });

  it('removes col', () => {
    // wrapper.update();
    instance.removeCol(1);
    // wrapper.update();
    expect(Helpers.removeItem).toHaveBeenCalledTimes(data.cells.length + 1)
    expect(Helpers.removeItem).toHaveBeenLastCalledWith(data.columnNames, 1)
    expect(wrapper.state('cells')).toEqual([
      ['a1', 'a3'],
      ['b1', 'b3'],
      ['c1', 'c3'],
    ]);
    expect(wrapper.state('columnNames')).toEqual(['col1', 'col3']);

    instance.removeCol(0);
    // wrapper.update();
    expect(Helpers.removeItem).toHaveBeenLastCalledWith(['col1', 'col3'], 0)
    expect(wrapper.state('cells')).toEqual([
      ['a3'],
      ['b3'],
      ['c3'],
    ]);
    expect(wrapper.state('columnNames')).toEqual(['col3']);
  });

  it('updates data cell', () => {
    // wrapper.update();
    const mockEvent = { target: { value: '45' } } as React.ChangeEvent<HTMLInputElement>
    instance.updateCell(1, 1, mockEvent);
    // wrapper.update();
    expect(Helpers.updateItem).toHaveBeenCalledTimes(2)
    expect(Helpers.updateItem).toHaveBeenLastCalledWith(data.cells, 1, ['b1', '45', 'b3'])
    expect(wrapper.state('cells')).toEqual([
      ['a1', 'a2', 'a3'],
      ['b1', '45', 'b3'],
      ['c1', 'c2', 'c3'],
    ]);
  });

  it('updates column name', () => {
    // wrapper.update();
    const mockEvent = { target: { value: 'New name' } } as React.ChangeEvent<HTMLInputElement>
    instance.updateColName(1,  mockEvent);
    expect(Helpers.updateItem).toHaveBeenCalledTimes(1)
    expect(Helpers.updateItem).toHaveBeenLastCalledWith(data.columnNames, 1, 'New name')
    expect(wrapper.state('columnNames')).toEqual(['col1', 'New name', 'col3']);
  });

  it('calls onChange prop', () => {
    onChange.mockClear();
    wrapper = shallow(<Table onChange={onChange} data={data} />);
    // adding column
    instance.addCol(0);
    // wrapper.update();
    const cellsWithNewRow = [
      ['', 'a1', 'a2', 'a3'],
      ['', 'b1', 'b2', 'b3'],
      ['', 'c1', 'c2', 'c3'],
    ]
    expect(onChange).toHaveBeenCalledWith({
      columnNames: ['New column 1', ...data.columnNames],
      cells: cellsWithNewRow
    });
    // adding row
    onChange.mockClear();
    instance.addRow(0);
    // wrapper.update();
    expect(onChange).toHaveBeenCalledWith({
      columnNames: ['New column 1', ...data.columnNames],
      cells: [
        ['', '', '', ''],
        ['', 'a1', 'a2', 'a3'],
        ['', 'b1', 'b2', 'b3'],
        ['', 'c1', 'c2', 'c3'],
      ],
    });
    // updating cell
    onChange.mockClear();
    const mockEvent = { target: { value: '22' } } as React.ChangeEvent<HTMLInputElement>
    instance.updateCell(0, 0, mockEvent);
    // wrapper.update();
    expect(onChange).toHaveBeenCalledWith({
      columnNames: ['New column 1', ...data.columnNames],
      cells:[
        ['22', '', '', ''],
        ['', 'a1', 'a2', 'a3'],
        ['', 'b1', 'b2', 'b3'],
        ['', 'c1', 'c2', 'c3'],
      ],
    });

    // TODO: more onChange tests (remove row, remove col, etc)
  });

  it('shows hints if no data', () => {
    const localWrapper: ShallowWrapper<TableProps, TableState, Table> = shallow(
      <Table
        onChange={onChange}
        data={{
          columnNames: [],
          cells: []
        }}
      />
    );

    expect(localWrapper.find(Styled.Hint).first().dive().find('p')).toHaveLength(1)
    expect(localWrapper.find(Styled.Hint).first().text()).toBe('No columns defined.')

    expect(localWrapper.find(Styled.Hint).at(1).dive().find('p')).toHaveLength(1)
    expect(localWrapper.find(Styled.Hint).at(1).text()).toBe('No rows defined.')
  });

  it('does not show hints if data defined', () => {
    expect(wrapper.find(Styled.Hint).first().text()).toBe('')
    expect(wrapper.find(Styled.Hint).at(1).text()).toBe('')
  });
});
