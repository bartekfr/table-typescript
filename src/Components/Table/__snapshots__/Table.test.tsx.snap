// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component matches initial snapshot 1`] = `
<styled.div>
  <table>
    <tbody>
      <tr>
        <th>
          <MenuCol
            add={[Function]}
            index={-1}
          />
          <MenuRow
            add={[Function]}
            index={-1}
          />
          id
        </th>
        <th
          key="0"
        >
          <MenuCol
            add={[Function]}
            index={0}
            remove={[Function]}
          />
          <CellInput
            onChange={[Function]}
            value="col1"
          />
        </th>
        <th
          key="1"
        >
          <MenuCol
            add={[Function]}
            index={1}
            remove={[Function]}
          />
          <CellInput
            onChange={[Function]}
            value="col2"
          />
        </th>
        <th
          key="2"
        >
          <MenuCol
            add={[Function]}
            index={2}
            remove={[Function]}
          />
          <CellInput
            onChange={[Function]}
            value="col3"
          />
        </th>
      </tr>
      <tr
        key="0"
      >
        <th>
          <MenuRow
            add={[Function]}
            index={0}
            remove={[Function]}
          />
          1
        </th>
        <td
          key="0"
        >
          <CellInput
            onChange={[Function]}
            value="a1"
          />
        </td>
        <td
          key="1"
        >
          <CellInput
            onChange={[Function]}
            value="a2"
          />
        </td>
        <td
          key="2"
        >
          <CellInput
            onChange={[Function]}
            value="a3"
          />
        </td>
      </tr>
      <tr
        key="1"
      >
        <th>
          <MenuRow
            add={[Function]}
            index={1}
            remove={[Function]}
          />
          2
        </th>
        <td
          key="0"
        >
          <CellInput
            onChange={[Function]}
            value="b1"
          />
        </td>
        <td
          key="1"
        >
          <CellInput
            onChange={[Function]}
            value="b2"
          />
        </td>
        <td
          key="2"
        >
          <CellInput
            onChange={[Function]}
            value="b3"
          />
        </td>
      </tr>
      <tr
        key="2"
      >
        <th>
          <MenuRow
            add={[Function]}
            index={2}
            remove={[Function]}
          />
          3
        </th>
        <td
          key="0"
        >
          <CellInput
            onChange={[Function]}
            value="c1"
          />
        </td>
        <td
          key="1"
        >
          <CellInput
            onChange={[Function]}
            value="c2"
          />
        </td>
        <td
          key="2"
        >
          <CellInput
            onChange={[Function]}
            value="c3"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <styled.p
    alert={true}
  />
  <styled.p
    alert={true}
  />
  <ExportCSV
    data={
      Object {
        "data": Array [
          Array [
            "a1",
            "a2",
            "a3",
          ],
          Array [
            "b1",
            "b2",
            "b3",
          ],
          Array [
            "c1",
            "c2",
            "c3",
          ],
        ],
        "fields": Array [
          "col1",
          "col2",
          "col3",
        ],
      }
    }
  />
</styled.div>
`;
